File: ios/Classes/PlacementChannelManager.swift
import Flutter
import UnityAds

public class PlacementChannelManager {
    
    var placementChannels: [String: FlutterMethodChannel];
    let binaryMessenger: FlutterBinaryMessenger
    
    init(binaryMessenger: FlutterBinaryMessenger) {
        self.binaryMessenger = binaryMessenger
        self.placementChannels = [String: FlutterMethodChannel]()
    }

    public func invokeMethod(_ methodName: String, _ placementId: String) {
        invokeMethod(methodName, placementId, arguments: [:])
    }

    public func invokeMethod(_ methodName: String, _ placementId: String, withErrorCode code: String, withMessage message: String) {
        var arguments: [String:String] = [:]
        arguments[UnityAdsConstants.ERROR_CODE_PARAMETER] = code
        arguments[UnityAdsConstants.ERROR_MESSAGE_PARAMETER] = message
        invokeMethod(methodName, placementId, arguments: arguments)
    }

    func invokeMethod(_ methodName: String,_ placementId: String, arguments: [String:String]){
        var args = arguments;
        args[UnityAdsConstants.PLACEMENT_ID_PARAMETER]=placementId
        var channel = placementChannels[placementId]
        if (channel == nil) {
            channel = FlutterMethodChannel(name: UnityAdsConstants.VIDEO_AD_CHANNEL + "_" + placementId, binaryMessenger: binaryMessenger)
            placementChannels[placementId] = channel
        }
        channel?.invokeMethod(methodName, arguments: args)
    }
}

File: android/src/main/java/com/rebeloid/unity_ads/PlacementChannelManager.java
package com.rebeloid.unity_ads;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodChannel;

public class PlacementChannelManager {
    private final Map<String, MethodChannel> placementChannels;
    private final BinaryMessenger binaryMessenger;

    public PlacementChannelManager(BinaryMessenger binaryMessenger) {
        this.placementChannels = new HashMap<>();
        this.binaryMessenger = binaryMessenger;
    }

    public void invokeMethod(String methodName, String adUnitId) {
        invokeMethod(methodName, adUnitId, new HashMap<>());
    }

    public void invokeMethod(String methodName, String adUnitId, String errorCode, String errorMessage) {
        Map<String, String> arguments = new HashMap<>();
        arguments.put(UnityAdsConstants.ERROR_CODE_PARAMETER, errorCode);
        arguments.put(UnityAdsConstants.ERROR_MESSAGE_PARAMETER, errorMessage);
        invokeMethod(methodName, adUnitId, arguments);
    }

    public void invokeMethod(String methodName, String placementId, Map<String, String> arguments) {
        arguments.put(UnityAdsConstants.PLACEMENT_ID_PARAMETER, placementId);
        MethodChannel channel = placementChannels.get(placementId);
        if (channel == null) {
            channel = new MethodChannel(binaryMessenger, UnityAdsConstants.VIDEO_AD_CHANNEL + "_" + placementId);
            placementChannels.put(placementId, channel);
        }
        channel.invokeMethod(methodName, arguments);
    }
}
